core/domain/prod_validation_jobs_one_off.py

```diff
@@ -321,39 +338,50 @@ def _get_external_model_properties(cls):
         This should be implemented by subclasses.
 
         Returns:
-            tuple(str, list(tuple), dict). A tuple with first element as
-                external model name, second element as a tuple of
-                cls.external_instance_details and the third element
-                as a properties dict with key as property name in summary
-                model and value as property name in external model.
+            tuple(str, str, dict). A tuple with first element as
+                external model name, second element as a key to fetch
+                external model details from cls.external_instance_details
+                and the third element as a properties dict with key as
+                property name in summary model and value as property name
+                in external model.
 
         Raises:
             NotImplementedError. This function has not yet been implemented.
         """
         raise NotImplementedError
 
     @classmethod
-    def _validate_external_model_properties(cls, item):
+    def _validate_external_model_properties(
+            cls, item, external_instance_details):
         """Validate that properties of the model match the corresponding
         properties of the external model.
 
         Args:
             item: ndb.Model. BaseSummaryModel to validate.
+            external_instance_details: dict(str, (list(str, str, ndb.Model))).
+                A dict is keyed by field name. Each value consists of a list
```

Comment:
TBH, I'm actually finding it quite hard to keep track of this dict, since this sort of custom handmade object is rather hard to parse.

Do you think it would be OK to define a domain class to represent the each of the individual values? Something like ExternalModelReference perhaps. And then we can have reference.model_name, reference.model_field_name, reference.properties (with a separate domain object for the properties dict). Or maybe the properties can be split up and made part of the ExternalModelReference object, that way you avoid complicated chains like reference.properties.XXX. What do you think?

For the outer dict/list layer, please explain also, in the docstring, what is meant by "field name" and how each value relates to the corresponding field name. The docstring talks about the structure but doesn't really convey anything about the meaning.

(Also, if you like, maybe we can iterate on just this one and then you'll only make changes to the whole file once we've settled on something? Don't want to waste your time making changes to the entire file and then having to do that repeatedly.)

Reply:
Makes sense, I agree it is confusing to understand what's happening here. I have introduced the domain object change for the `BaseValidators` and `ActivityReferencesModelValidator ` [here](https://github.com/oppia/oppia/pull/9495/commits/13f1a4deeaaf919f47cfbf90b57c8226275cfcd1). PTAL, Thanks!